src/main/java/com/ita/if103java/ims/dao/impl/WarehouseDaoImpl.java

```diff
@@ -0,0 +1,167 @@
+    public Warehouse findById ( long id){
+        try {
+            return jdbcTemplate.queryForObject(QueryToWarehouse.SQL_SELECT_WAREHOUSE_BY_ID, warehouseRowMapper, id);
+        } catch (EmptyResultDataAccessException e) {
+            throw new EntityNotFoundException(e.getMessage(), "id = " + id);
+        } catch (DataAccessException e) {
+            throw crudException(e.getMessage(), "get", "id = " + id);
+        }
+
+        }
+
+    @Override
+    public List<Warehouse> findByParent(long id) {
+        return null;
+    }
+
+    @Override
+    public Warehouse create(Warehouse warehouse) {
+        try {
+            GeneratedKeyHolder holder = new GeneratedKeyHolder();
+            jdbcTemplate.update(connection -> createWarehouseStatement(warehouse, connection), holder);
+
+            warehouse.setId(Optional.ofNullable(holder.getKey())
+                .map(Number::longValue)
+                .orElseThrow(() -> new CRUDException("Error during an user creation: " +
+                    "Autogenerated key is null")));
+
+
+        } catch (DataAccessException e) {
+            throw crudException(e.getMessage(), "create", "id = " + warehouse.getId());
+        }
+        return warehouse;
+    }
+
+
+
+    @Override
+        public Warehouse update (Warehouse warehouse){
+        try{
+        int status = 0;
+        status = jdbcTemplate.update(QueryToWarehouse.SQL_UPDATE_WAREHOUSE,
+            warehouse.getName(), warehouse.getInfo(), warehouse.getCapacity(),
+            warehouse.isBottom(),warehouse.getParentID(),warehouse.getAccountID(),
+            warehouse.isActive(),warehouse.getChildren());
+
+         }catch (DataAccessException e) {
+            throw crudException(e.getMessage(), "update", "id = " + warehouse.getId());
+        }
+
+        return warehouse;
+        }
+
+        @Override
+        public boolean deleteWarehouse (long id){
+        int status;
+        try{
+            status = jdbcTemplate.update(QueryToWarehouse.SQL_DELETE_WAREHOUSE, false,id);
+        }
+        catch (DataAccessException e){
+            throw crudException(e.getMessage(), "delete", "id = " + id);
+        }if (status == 0)
+            throw warehouseEntityNotFoundException("Delete warehouse exception ", "id " + id);
+
+            return true;
+        }
+
+    private EntityNotFoundException warehouseEntityNotFoundException(String message, String attribute) {
+        EntityNotFoundException exception = new EntityNotFoundException(message);
+        LOGGER.error("EntityNotFoundException exception. Warehouse is not found ({}). Message: {}", attribute, message);
+        return exception;
+    }
+
+    private CRUDException crudException(String message, String operation, String attribute) {
+        CRUDException exception = new CRUDException(message);
+        LOGGER.error("CRUDException exception. Operation:({}) warehouse ({}) exception. Message: {}", operation, attribute, message);
+        return exception;
+    }
```

Comment:
Please sit together with the team and think how to move this similar methods to some abstract class and generalize them in order they can be access by every DAO